### PR TITLE
Refactor relation creation: introduce RelationDef

### PR DESCRIPTION
**Pull request by CodeLionX**
## Proposed Changes

  - new `RelationDef` trait for defining relations
  - create actual store with `RowRelation(relationDef)`
  - example:

```scala
object TestRelation extends RelationDef {
  val col1: ColumnDef[Int] = ColumnDef("col1")
  val col2: ColumnDef[String] = ColumnDef("col2")

  override val columns: Set[UntypedColumnDef] = Set(col1, col2)
  override val name: String = "testRelation"
}

// `RowRelation` can be replaced by any other relational store
val testRelation: MutableRelation = RowRelation(TestRelation)
testRelation.insert(
  TestRelation.newRecord.build(), // preferred
  testRelation.newRecord.build()
)
val records: Relation = testRelation.where(TestRelation.col1 -> { _ == 1 })
```

## Related

  -

### CODE REVIEW

**Comment by CodeLionX** [comment]
**Naming Conventions:**
- rename `RowRelation` to `RowStore`?

Change current
- definition: `UserDef`
- actual instance: `User`
to
- definition: `User`
- actual instance: `user`

File: adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/DactorTest.scala [hunk 1]
```diff
@@ -31,22 +31,19 @@ object DactorTest {
...
+    object TestRelationDef extends RelationDef {
       val col1: ColumnDef[Int] = ColumnDef("col1")
       val col2: ColumnDef[String] = ColumnDef("col2")
-      val columns: Set[UntypedColumnDef] = Set(col1, col2)
 
-      def rowRelation: MutableRelation = new RowRelation {
-        override val columns: Set[UntypedColumnDef] = TestRelationDef.columns
-      }
+      override val columns: Set[UntypedColumnDef] = Set(col1, col2)
+      override val name: String = "testrelation"
```
Comment (srfc) [inline adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/DactorTest.scala]: Why is the name here `testrelation` while it is known to its `Dactor` as `testRelation`?
And does a `RelationDef` really need a name?

**Comment by srfc** [comment]
I'm actually not sure if I like the `SomethingDef` naming for `RelationDef`s, because when interacting with the `Relation`, e.g. creating a new `Record` and using the `Relation`s `ColumnDef`s, I feel like using `RelationDef.ColumnName` adds another layer of indirection that I feel is unnecessary. This does not relate to the `trait` itself as much as the naming of all `RelationDef`s.

I think I would prefer that, if everything stays the same, the suggested pattern would be
```scala
// companion object
object UserInfo extends RelationDef {
  // ...
}

// corresponding Dactor - as is defined in the PR now:
val userInfo: MutableRelation = RowRelation(UserInfo)

override protected val relations: Map[String, MutableRelation] = Map("userInfo" -> userRelation)

override def receive: Receive = Actor.emptyBehavior
```

> rename RowRelation to RowStore?

I think this only really makes sense when we decide to call the technical side of things `Store` and the definitions "go back" to be called `Relation` without the `Def` only. What do you think abou this? Otherwise I think the naming can remain as is.

File: adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/DactorTest.scala (hunk 1 above)
Comment (CodeLionX) [reply adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/DactorTest.scala]: it's just a test and I haven't thought about it very much :laughing: 
I will change this.

I already thought of introducing a declared name, which you can later use to send the insert messages to. Like I did in `DactorTest`:

```scala
"insert matching record successfully" in {
      val insertMessage = DefaultMessagingProtocol.InsertIntoRelation(
        TestRelationDef.name,     // <-- here
        Seq(
          Record(TestRelationDef.columns)(
            TestRelationDef.col1 ~> 1 &
            TestRelationDef.col2 ~> "1"
          ).build()
      ))
      dut.tell(insertMessage, probe.ref)
      probe.expectMsg(akka.actor.Status.Success)
}
```

I can imagine to let the `Dactor` handle the naming of its relations, as it is currently the only one using it:

```scala
class MyDactor(id: Int) extends Dactor(id) {
  val firstRel = RowRelation(RelDef1)
  val secondRel = RowRelation(RelDef2)

  override val declareRelations: Set[MutableRelation] = Set(firstRel, secondRel)

  // in `Dactor` trait:
  val relations: Map[String, MutableRelation] = declareRelations.map{ rel =>
    rel.name -> rel
  }.toMap
}
```
Then you are sure, that you can send messages to a `Dactor` and specify the relation by its name `DactorComp.RelDef.name`.

**Comment by CodeLionX** [comment]
@srfc I addressed your remarks, I hope this is now ok for you?

**Pull request merged**